---
title: "Class 12: Transcriptomics and the analysis of RNA-Seq data"
author: "Rogelio Castro"
format: pdf
editor: visual
---

Here we will use the DESeq2 package for RNASeq analysis. The data for today's class comes from a study of airway smooth muscle cells were treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al. 2014).


## Import their data

WE need two things for this analysis:

-countData(Counts for every transcript/gene in each experiment)
-colData (metadata that describes the experimental setup)

```{r}
countData <- read.csv("airway_scaledcounts.csv",row.names = 1 )
head(countData)
```

```{r}
metadata <- read.csv("airway_metadata.csv")
metadata
```


>Q1. How many genes are in this dataset? 

```{r}
nrow(countData)
```

>Q2. How many ‘control’ cell lines do we have?

```{r}
table(metadata$dex)
```
Another way

```{r}
sum(metadata$dex == "control")
```

-Step 1.
Calculate the mean of the control samples (i.e. columns in countData)
Calculate the mean of the treated samples

(a) We need to find which columns are "control" samples. 
- look in the metadata (a.k.a. colData), $dex column

```{r}
control.inds <- metadata$dex == "control"
control.inds
```

(b) Extract all the control columns from `countData` and call it `control.counts`

```{r}
control.counts <- countData[ , control.inds]
head(control.counts)
```

(c) Calculate the eman value across the rows of `control.counts` i.e. calculate the mean count values for each gene in the control samples.

```{r}
control.means <- rowMeans(control.counts)
head(control.means)
```

-Step 2.
Calculate the mean of the treated samples...

```{r}
treated.inds <- metadata$dex == "treated"
treated.inds
```

```{r}
treated.counts <- countData[ , treated.inds]
head(treated.counts)
```

```{r}
treated.means <- rowMeans(treated.counts)
head(treated.means)
```

We now have control and treated mean count values. For ease of book-keeping I will combine these vectors into a new data.frame called `meancounts`.

```{r}
meancounts <- data.frame (control.means, treated.means)
head(meancounts)
```

```{r}
plot(meancounts)
```

```{r}
plot(meancounts, log="xy")
```


We use log transofrms fror skewed data such as this and because we really care most about realtive changes in magnitude.

We mos often use log2 as our transform as the math is easier to interpret than log10 or others.

If we have no change -i.e. same values in control and treated we will have a log2 value of zero.

```{r}
log2(20/20)
```

If ia have dpuble the amount i.e. 20 compated to the 10 for example. I will have a log2 fold-change of +1
```{r}
log2(20/10)
```

If I have half the amount I will have a log2 fold-change of -1
```{r}
log2(10/20)
```

```{r}
log2(40/10)
```

```{r}
meancounts$log2fc <- log2(meancounts$treated.means / meancounts$control.means)
head(meancounts)
```

>Q. How many genes are up regulated at the common treshold of +2 log2FC values?

```{r}
sum(meancounts$log2fc >= 2, na.rm = TRUE)
```

Hold on what about the stats! Yes, these are big changes bit are these changes significant?!
 
To do this properly, we will turn to the DESeq2 package.


## DESeq2 analysis:

```{r}
#| message: false
library(DESeq2)
```

To use DESeq we need our input contData and colData in a specific format that DESeq wants:

```{r}
dds <- DESeqDataSetFromMatrix(countData = countData,
                       colData = metadata,
                       design = ~dex)

```

To run analysis, I can now use the main DESeq2 function called `DESeq()` with `dds` as input.

```{r}
dds <- DESeq(dds)
```

To get the results out of. this `dds` object we can use the `results()` function from the package.

```{r}
res <- results(dds)
head(res)
```

## Volcano plot:
Let's make a final (for today) plot of log2 fold-change vs the adjusted P-value. 

```{r}
plot(res$log2FoldChange, res$padj)
```

It is the low P-values that we care about and these are lost in the skewed plot above. Let's take the log of the $padj values for our plot.

```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(v=c(+2,-2), col="blue")
abline(h=log(0.05), col="blue")
```

Finally we can make a color vector to use in the plot to better highlight the genes we care about.

```{r}
mycols <- rep("gray", nrow(res))
mycols[abs(res$log2FoldChange) >= 2] <- "red"
mycols[res$padj >0.05 ] <- "gray"

plot(res$log2FoldChange, -log(res$padj), col=mycols)
abline(v=c(+2,-2), col="blue")
abline(h=log(0.05), col="blue")

```

















